File: public/js/bim-registry.js
import { Box3, Matrix4, Vector3 } from 'three';

// Rendering batches are deliberately absent from this index. Physical counts
// and identity survive visibility, clipping, material changes and batch rebuilds.
export function parseBimRegistry(data, meshes, catalog) {
  const require = (ok, message) => { if (!ok) throw new Error(`BIM registry: ${message}`); };
  const finite = Number.isFinite;
  const index = (rows, label) => {
    require(Array.isArray(rows), `${label} must be an array`);
    const result = new Map();
    for (const row of rows) {
      require(row && typeof row.id === 'string' && row.id.length > 0 && row.id.length <= 256 && !result.has(row.id), `${label} missing or duplicate ID`);
      result.set(row.id, row);
    }
    return result;
  };
  const matrix = values => {
    require(Array.isArray(values) && values.length === 16 && values.every(finite), 'invalid matrix');
    require([3, 7, 11].every(i => Math.abs(values[i]) < 1e-9) && Math.abs(values[15] - 1) < 1e-9, 'non-affine matrix');
    const m = new Matrix4().fromArray(values);
    require(Math.abs(m.determinant()) > 1e-12, 'singular matrix');
    return m;
  };
  require(data?.schemaVersion === 1 && data.units === 'm' && data.coordinateSystem === 'glTF-Y-up', 'unsupported format');
  require(data.modelId === catalog.id && /^[0-9a-f]{64}$/.test(data.sourceSha256) && data.sourceSha256 === catalog.sourceSha256, 'wrong model revision');
  require(finite(data.conversionTolerance) && data.conversionTolerance > 0 && data.conversionTolerance <= 0.001, 'invalid tolerance');
  const elements = index(data.elements, 'elements'), types = index(data.types, 'types'), families = index(data.families, 'families');
  const storeys = index(data.storeys, 'storeys'), openings = index(data.openings, 'openings');
  const allIds = [...elements.keys(), ...types.keys(), ...families.keys(), ...storeys.keys(), ...openings.keys()];
  require(new Set(allIds).size === allIds.length, 'ID namespace collision');
  require(elements.size && types.size && storeys.size, 'empty registry');
  for (const s of storeys.values()) require(finite(s.elevation), 'invalid storey elevation');
  for (const t of types.values()) require(families.has(t.familyId) && elements.get(t.prototypeElementId)?.typeId === t.id && Number.isInteger(t.revision) && t.revision > 0, 'unresolved type reference');
  const source = new Map();
  for (const mesh of meshes) {
    const id = mesh.userData.viewer_id;
    require(typeof id === 'string' && !source.has(id), 'missing or duplicate source component'); source.set(id, mesh);
  }
  const owners = new Map(), members = new Map(), bounds = new Map();
  const actual = new Matrix4(), a = new Vector3(), b = new Vector3();
  for (const e of elements.values()) {
    const type = types.get(e.typeId);
    require(type && families.get(type.familyId)?.category === e.category, 'unresolved type or category mismatch');
    require(e.membership === 'reviewed' && storeys.has(e.primaryStorey), 'unreviewed membership or unknown storey');
    require(Array.isArray(e.storeyRefs) && new Set(e.storeyRefs).size === e.storeyRefs.length && e.storeyRefs.every(id => id !== e.primaryStorey && storeys.has(id)), 'invalid spatial references');
    require(Array.isArray(e.roomIds) && e.roomIds.every(id => typeof id === 'string'), 'invalid room references');
    require(e.evidence && ['authored', 'inferred'].includes(e.evidence.basis) && ['low', 'medium', 'high'].includes(e.evidence.confidence) && typeof e.evidence.source === 'string', 'missing evidence');
    require(Array.isArray(e.bounds) && e.bounds.length === 2 && e.bounds.every(p => Array.isArray(p) && p.length === 3 && p.every(finite)) && e.bounds[0].every((x, i) => x <= e.bounds[1][i]), 'invalid bounds');
    const root = matrix(e.rootMatrix), list = [], slots = new Set();
    require(Array.isArray(e.components) && e.components.length, 'empty assembly');
    for (const c of e.components) {
      require(!owners.has(c.id) && source.has(c.id), 'missing component or multiple product owners');
      require(Number.isInteger(c.slot) && c.slot >= 0 && !slots.has(c.slot), 'invalid component slot'); slots.add(c.slot);
      const mesh = source.get(c.id); actual.multiplyMatrices(root, matrix(c.localMatrix));
      if (!mesh.geometry.boundingBox) mesh.geometry.computeBoundingBox();
      const box = mesh.geometry.boundingBox;
      for (let corner = 0; corner < 8; corner++) {
        a.set(corner & 1 ? box.max.x : box.min.x, corner & 2 ? box.max.y : box.min.y, corner & 4 ? box.max.z : box.min.z);
        b.copy(a).applyMatrix4(mesh.matrixWorld); a.applyMatrix4(actual);
        require(a.distanceTo(b) <= data.conversionTolerance, `transform mismatch for ${c.id}`);
      }
      owners.set(c.id, e); list.push(mesh);
    }
    for (const [name, q] of Object.entries(e.quantities || {})) {
      require(['openingArea', 'outerFrameArea', 'netGlazingArea', 'floorArea'].includes(name), 'unknown quantity');
      if (q !== null) require(q && finite(q.value) && q.value >= 0 && q.unit === 'm2' && typeof q.method === 'string' && typeof q.source === 'string', 'invalid typed area');
    }
    let ancestor = e; const chain = new Set([e.id]);
    while (ancestor.parentId) {
      require(elements.has(ancestor.parentId) && !chain.has(ancestor.parentId), 'cyclic or unresolved parent');
      chain.add(ancestor.parentId); ancestor = elements.get(ancestor.parentId);
    }
    members.set(e.id, list); bounds.set(e.id, new Box3(new Vector3().fromArray(e.bounds[0]), new Vector3().fromArray(e.bounds[1])));
  }
  const unresolved = index(data.unresolvedComponents, 'unresolved components');
  require(owners.size + unresolved.size === source.size && [...unresolved.keys()].every(id => source.has(id) && !owners.has(id)), 'incomplete component coverage');
  const fillingIds = new Set();
  for (const o of openings.values()) {
    require(elements.get(o.hostId)?.ifcClass === 'IfcWall' && ['IfcWindow', 'IfcDoor'].includes(elements.get(o.fillingId)?.ifcClass) && elements.get(o.fillingId)?.openingId === o.id && !fillingIds.has(o.fillingId), 'unresolved/ambiguous opening');
    fillingIds.add(o.fillingId); matrix(o.rootMatrix);
    require(finite(o.depth) && o.depth > 0 && Array.isArray(o.profile) && o.profile.length >= 3 && o.profile.every(p => Array.isArray(p) && p.length === 2 && p.every(finite)), 'invalid opening profile');
  }
  for (const e of elements.values()) require(!e.openingId || openings.has(e.openingId), 'unresolved product opening');
  require(Array.isArray(data.coverage), 'missing coverage');
  const coverage = new Map();
  for (const c of data.coverage) {
    require(c && typeof c.category === 'string' && !coverage.has(c.category) && ['partial', 'complete-modeled-category'].includes(c.status) && typeof c.scope === 'string', 'invalid coverage');
    require(c.status !== 'complete-modeled-category' || ![...unresolved.values()].some(u => u.category === c.category), 'complete category still has unresolved components');
    coverage.set(c.category, c);
  }
  for (const e of elements.values()) require(coverage.has(e.category), 'missing category coverage');
  const counts = [...coverage.values()].map(c => {
    const products = [...elements.values()].filter(e => e.category === c.category);
    const areas = {};
    for (const key of ['openingArea', 'outerFrameArea', 'netGlazingArea', 'floorArea']) {
      const known = products.filter(e => e.quantities?.[key] !== null && e.quantities?.[key] !== undefined);
      areas[key] = { value: known.length ? known.reduce((n, e) => n + e.quantities[key].value, 0) : null, unit: 'm2', known: known.length, registered: products.length };
    }
    return { category: c.category, count: products.length, status: c.status, scope: c.scope, areas };
  });
  const representatives = [...elements.values()].map(e => members.get(e.id)[0]).concat([...unresolved.keys()].map(id => source.get(id)));
  return { data, elements, types, families, counts, representatives,
    product: mesh => owners.get(mesh.userData.viewer_id) || null,
    members: mesh => members.get(owners.get(mesh.userData.viewer_id)?.id) || [mesh],
    bounds: mesh => bounds.get(owners.get(mesh.userData.viewer_id)?.id) || mesh.userData.bounds,
  };
}

export function describeProduct(mesh, registry) {
  const e = registry?.product(mesh); if (!e) return null;
  const type = registry.types.get(e.typeId), family = registry.families.get(type.familyId);
  const prop = (label, value, basis = 'authored', source = e.evidence.source) => ({ label, value: String(value), basis, source, confidence: e.evidence.confidence });
  const label = key => ({ stationId: 'Station ID', seatNumber: 'Seat number', seatFunction: 'Seat function', openingArea: 'Opening area', outerFrameArea: 'Outer-frame projected area', netGlazingArea: 'Net glazing area', floorArea: 'Modeled floor footprint', leafCount: 'Door leaves' })[key] || key.replace(/([A-Z])/g, ' $1').toLowerCase().replace(/^./, c => c.toUpperCase());
  const size = new Vector3(); registry.bounds(mesh).getSize(size);
  const properties = [prop('Category', e.category), prop('Family', family.name), prop('Type', type.name),
    prop('Product type ID', type.id), prop('Type revision', type.revision), prop('Assembly', `${e.components.length} components · one modeled product`),
    prop('Primary storey', e.primaryStorey), prop('Rooms', e.roomIds.join(', ') || 'Not assigned'),
    prop('Model extent · X × Y × Z', size.toArray().map(n => n.toFixed(2)).join(' × ') + ' m', 'measured', 'Whole-product world bounding box; extent is not opening area')];
  for (const [key, value] of Object.entries(e.properties || {})) properties.push(prop(label(key), value));
  for (const [key, q] of Object.entries(e.quantities || {})) properties.push(prop(label(key), q ? `${q.value.toFixed(3)} m²` : 'Unknown', q ? 'measured' : 'authored', q ? `${q.method}. ${q.source}` : 'No reviewed measurement supplied'));
  return { title: e.name, sourceName: mesh.name, id: e.id, properties };
}

File: scripts/bim_registry.py
"""Semantic and transform validation, with no optional BIM dependencies.

The JSON schema supplies additional shape validation in the offline release
check. This validator also runs during normal model imports.
"""
import math,re,uuid

def finite(value):
    return isinstance(value,(int,float)) and not isinstance(value,bool) and math.isfinite(value)

def identity(): return [1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1]

def multiply(a,b):
    return [sum(a[k*4+r]*b[c*4+k] for k in range(4)) for c in range(4) for r in range(4)]

def transform(a,p):
    return [sum(a[k*4+r]*p[k] for k in range(3))+a[12+r] for r in range(3)]

def node_matrix(n):
    if 'matrix' in n:return n['matrix']
    x,y,z,w=n.get('rotation',[0,0,0,1]);s=n.get('scale',[1,1,1]);t=n.get('translation',[0,0,0])
    return [(1-2*(y*y+z*z))*s[0],2*(x*y+z*w)*s[0],2*(x*z-y*w)*s[0],0,
            2*(x*y-z*w)*s[1],(1-2*(x*x+z*z))*s[1],2*(y*z+x*w)*s[1],0,
            2*(x*z+y*w)*s[2],2*(y*z-x*w)*s[2],(1-2*(x*x+y*y))*s[2],0,*t,1]

def document_nodes(doc):
    nodes={}; seen=set()
    def visit(i,parent):
        if i in seen:raise ValueError('Cyclic/multiple node parent')
        seen.add(i);n=doc['nodes'][i];world=multiply(parent,node_matrix(n))
        if 'mesh' in n:
            ident=n.get('extras',{}).get('viewer_id')
            if not ident or ident in nodes:raise ValueError('Missing/duplicate source component ID')
            nodes[ident]={'world':world,'node':n}
        for child in n.get('children',[]):visit(child,world)
    for i in doc['scenes'][doc.get('scene',0)]['nodes']:visit(i,identity())
    return nodes

def validate_registry(data, doc=None, model_id=None, source_hash=None):
    def require(ok,message):
        if not ok:raise ValueError('BIM registry: '+message)
    def index(rows,label):
        require(isinstance(rows,list),label+' must be an array')
        result={}
        for row in rows:
            require(isinstance(row,dict) and isinstance(row.get('id'),str) and 0<len(row['id'])<=256,label+' invalid ID')
            require(row['id'] not in result,label+' duplicate ID '+row['id']);result[row['id']]=row
        return result
    def mat(m):
        require(isinstance(m,list) and len(m)==16 and all(finite(x) for x in m),'nonfinite or malformed matrix')
        require(max(abs(m[i]-n) for i,n in [(3,0),(7,0),(11,0),(15,1)])<1e-9,'nonaffine matrix')
        det=m[0]*(m[5]*m[10]-m[9]*m[6])-m[4]*(m[1]*m[10]-m[9]*m[2])+m[8]*(m[1]*m[6]-m[5]*m[2])
        require(abs(det)>1e-12,'singular matrix')
    require(isinstance(data,dict) and data.get('schemaVersion')==1,'unsupported schema version')
    require(data.get('units')=='m' and data.get('coordinateSystem')=='glTF-Y-up','unexpected coordinate frame')
    require(isinstance(data.get('sourceSha256'),str) and re.fullmatch('[0-9a-f]{64}',data['sourceSha256']),'missing source SHA256')
    require(model_id is None or model_id==data.get('modelId'),'wrong model version')
    require(source_hash is None or source_hash==data.get('sourceSha256'),'wrong source revision')
    try:uuid.UUID(data['projectNamespace'])
    except (KeyError,ValueError,TypeError,AttributeError):raise ValueError('BIM registry: invalid project namespace')
    tolerance=data.get('conversionTolerance');require(finite(tolerance) and 0<tolerance<=.001,'invalid tolerance')
    storeys=index(data.get('storeys'),'storeys');families=index(data.get('families'),'families');types=index(data.get('types'),'types');elements=index(data.get('elements'),'elements');openings=index(data.get('openings'),'openings')
    require(bool(storeys) and bool(types) and bool(elements),'empty product registry')
    all_ids=list(storeys)+list(families)+list(types)+list(elements)+list(openings)
    require(len(all_ids)==len(set(all_ids)),'ID namespace collision')
    for s in storeys.values():require(finite(s.get('elevation')),'invalid storey elevation')
    for t in types.values():
        require(t.get('familyId') in families and t.get('prototypeElementId') in elements,'unresolved type reference')
        require(elements[t['prototypeElementId']].get('typeId')==t['id'],'prototype belongs to a different type')
        require(isinstance(t.get('revision'),int) and not isinstance(t['revision'],bool) and t['revision']>0,'invalid type revision')
    owned={};maximum=0;nodes=document_nodes(doc) if doc else None
    for e in elements.values():
        require(e.get('typeId') in types,'unresolved product type')
        require(e.get('category')==families[types[e['typeId']]['familyId']].get('category'),'family/category mismatch')
        require(e.get('primaryStorey') in storeys,'unresolved primary storey')
        refs=e.get('storeyRefs',[]);require(isinstance(refs,list) and len(refs)==len(set(refs)) and all(r in storeys and r!=e['primaryStorey'] for r in refs),'invalid additional storey references')
        mat(e.get('rootMatrix'))
        b=e.get('bounds');require(isinstance(b,list) and len(b)==2 and all(isinstance(p,list) and len(p)==3 and all(finite(x) for x in p) for p in b) and all(b[0][k]<=b[1][k] for k in range(3)),'invalid bounds')
        require(e.get('membership')=='reviewed','unreviewed product may not be counted')
        require(e.get('ifcClass') in ['IfcFurniture','IfcWindow','IfcDoor','IfcColumn','IfcWall','IfcStair','IfcSlab','IfcRamp','IfcRailing','IfcBuildingElementProxy'],'unsupported IFC class')
        evidence=e.get('evidence',{});require(evidence.get('basis') in ['authored','inferred'] and evidence.get('confidence') in ['low','medium','high'] and bool(evidence.get('source')),'missing evidence')
        require(isinstance(e.get('components'),list) and bool(e['components']),'empty product membership')
        slots=set()
        for c in e['components']:
            ident=c.get('id');require(isinstance(ident,str) and ident not in owned,'component has multiple product owners')
            require(isinstance(c.get('slot'),int) and c['slot']>=0 and c['slot'] not in slots,'invalid/duplicate component slot');slots.add(c['slot'])
            owned[ident]=e['id'];mat(c.get('localMatrix'))
            if nodes is not None:
                require(ident in nodes,'missing source component '+ident)
                expected=nodes[ident]['world'];actual=multiply(e['rootMatrix'],c['localMatrix'])
                n=nodes[ident]['node']
                for p in doc['meshes'][n['mesh']]['primitives']:
                    a=doc['accessors'][p['attributes']['POSITION']]
                    require('min' in a and 'max' in a,'position bounds required')
                    for corner in range(8):
                        point=[a['max' if corner&(1<<k) else 'min'][k] for k in range(3)]
                        err=math.dist(transform(expected,point),transform(actual,point));maximum=max(maximum,err)
                        require(err<=tolerance,'component transform mismatch '+ident)
        for name,q in e.get('quantities',{}).items():
            require(name in ['openingArea','outerFrameArea','netGlazingArea','floorArea'],'unknown quantity')
            if q is not None:require(finite(q.get('value')) and q['value']>=0 and q.get('unit')=='m2' and bool(q.get('method')) and bool(q.get('source')),'invalid typed area')
        ancestor=e;chain=set()
        while ancestor.get('parentId'):
            parent=ancestor['parentId'];require(parent in elements and parent not in chain,'unresolved/cyclic product parent');chain.add(parent);ancestor=elements[parent]
    unresolved=index(data.get('unresolvedComponents'),'unresolved components')
    require(not set(unresolved)&set(owned),'owned component also unresolved')
    if nodes is not None:require(set(nodes)==set(owned)|set(unresolved),'source component coverage mismatch')
    fillings=set()
    for o in openings.values():
        require(o.get('hostId') in elements and o.get('fillingId') in elements,'unresolved opening host/filling')
        require(elements[o['hostId']]['ifcClass']=='IfcWall' and elements[o['fillingId']]['ifcClass'] in ['IfcDoor','IfcWindow'],'invalid opening relationship classes')
        require(elements[o['fillingId']].get('openingId')==o['id'] and o['fillingId'] not in fillings,'ambiguous opening filling');fillings.add(o['fillingId'])
        mat(o.get('rootMatrix'));profile=o.get('profile')
        require(isinstance(profile,list) and len(profile)>=3 and all(isinstance(p,list) and len(p)==2 and all(finite(x) for x in p) for p in profile),'invalid opening profile')
        area=sum(a[0]*b[1]-b[0]*a[1] for a,b in zip(profile,profile[1:]+profile[:1]))/2
        require(abs(area)>1e-9 and finite(o.get('depth')) and o['depth']>0,'degenerate opening')
    for e in elements.values():require('openingId' not in e or e['openingId'] in openings,'unresolved product opening')
    coverage=data.get('coverage');require(isinstance(coverage,list),'missing quantity coverage')
    cats=set()
    for c in coverage:
        require(c.get('category') not in cats and c.get('status') in ['complete-modeled-category','partial'] and bool(c.get('scope')),'invalid category coverage');cats.add(c['category'])
        if c['status']=='complete-modeled-category':require(not any(u.get('category')==c['category'] for u in unresolved.values()),'complete category still has unresolved components')
    require(all(e['category'] in cats for e in elements.values()),'missing product coverage category')
    return {'products':len(elements),'ownedComponents':len(owned),'unresolvedComponents':len(unresolved),'maximumTransformErrorM':maximum,'status':'PASS'}
